### ISSUE TITLE
FP return_and_then

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#14927]
### Summary

.

### Reproducer

I tried this code:

```rust
#![warn(clippy::return_and_then)]
#![allow(unused)]
use std::path::Path;
struct A {
    pub func: fn(check: bool, a: &Path, b: Option<&Path>),
}
const MY_A: A = A {
    func: |check, a, b| {
        if check {
            let _ = ();
        } else if let Some(parent) = b.and_then(|p| p.parent()) {
            let _ = ();
        }
    },
};

pub fn main() {}

```
suggestion:
```
warning: use the `?` operator instead of an `and_then` call
  --> src/main.rs:11:38
   |
11 |         } else if let Some(parent) = b.and_then(|p| p.parent()) {
   |                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#return_and_then
note: the lint level is defined here
  --> src/main.rs:1:9
   |
1  | #![warn(clippy::return_and_then)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^
help: try
   |
11 ~         } else if let Some(parent) = {
12 +             let p = b?;
13 +             p.parent()
14 ~         } {
   |
```

the suggested code:

```rust
#![warn(clippy::return_and_then)]
#![allow(unused)]
use std::path::Path;
struct A {
    pub func: fn(check: bool, a: &Path, b: Option<&Path>),
}
const MY_A: A = A {
    func: |check, a, b| {
        if check {
            let _ = ();
        } else if let Some(parent) = {
            let p = b?;
            p.parent()
        } {
            let _ = ();
        }
    },
};

pub fn main() {}

```

does not compile:
```
error[E0277]: the `?` operator can only be used in a closure that returns `Result` or `Option` (or another type that implements `FromResidual`)
  --> src/main.rs:12:22
   |
8  |     func: |check, a, b| {
   |           ------------- this function should return `Result` or `Option` to accept `?`
...
12 |             let p = b?;
   |                      ^ cannot use the `?` operator in a closure that returns `()`

For more information about this error, try `rustc --explain E0277`.
```

### Version

```text

```

**Comment by matthiaskrgr** [comment]
```
rustc 1.89.0-nightly (1bbd62e54 2025-05-29)
binary: rustc
commit-hash: 1bbd62e547ba5cc08ccb44c27def3d33195d2dd5
commit-date: 2025-05-29
host: x86_64-unknown-linux-gnu
release: 1.89.0-nightly
LLVM version: 20.1.5
```
